<a href="https://colab.research.google.com/github/jwross24/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [1]:
### YOUR CODE STARTS HERE
from scipy.stats import ttest_ind, ttest_ind_from_stats, ttest_rel

import pandas as pd
import numpy as np


!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2019-10-08 18:46:23--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2019-10-08 18:46:24 (127 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [2]:
# Load Data
df = pd.read_csv('house-votes-84.data', 
                 header=None,
                 names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa'])
print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [3]:
df = df.replace({'?':np.NaN, 'n':0, 'y':1})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
# Create separate dataframes for republican and democraft
rep = df[df.party == "republican"]
dem = df[df.party == "democrat"]

In [0]:
from scipy.stats import ttest_ind

In [8]:
df.columns.tolist()

['party',
 'handicapped-infants',
 'water-project',
 'budget',
 'physician-fee-freeze',
 'el-salvador-aid',
 'religious-groups',
 'anti-satellite-ban',
 'aid-to-contras',
 'mx-missile',
 'immigration',
 'synfuels',
 'education',
 'right-to-sue',
 'crime',
 'duty-free',
 'south-africa']

In [15]:
for col in df.columns[1:].tolist():
  result = ttest_ind(rep[col], dem[col], nan_policy="omit")
  print(f"Issue: {col}\nt-statistic = {result.statistic}\npvalue = {result.pvalue}\n")

Issue: handicapped-infants
t-statistic = -9.205264294809222
pvalue = 1.613440327937243e-18

Issue: water-project
t-statistic = 0.08896538137868286
pvalue = 0.9291556823993485

Issue: budget
t-statistic = -23.21277691701378
pvalue = 2.0703402795404463e-77

Issue: physician-fee-freeze
t-statistic = 49.36708157301406
pvalue = 1.994262314074344e-177

Issue: el-salvador-aid
t-statistic = 21.13669261173219
pvalue = 5.600520111729011e-68

Issue: religious-groups
t-statistic = 9.737575825219457
pvalue = 2.3936722520597287e-20

Issue: anti-satellite-ban
t-statistic = -12.526187929077842
pvalue = 8.521033017443867e-31

Issue: aid-to-contras
t-statistic = -18.052093200819733
pvalue = 2.82471841372357e-54

Issue: mx-missile
t-statistic = -16.437503268542994
pvalue = 5.03079265310811e-47

Issue: immigration
t-statistic = 1.7359117329695164
pvalue = 0.08330248490425066

Issue: synfuels
t-statistic = -8.293603989407588
pvalue = 1.5759322301054064e-15

Issue: education
t-statistic = 20.500685724563073

In [0]:
dem['aid-to-contras'].mean()

0.8288973384030418

In [0]:
rep['aid-to-contras'].mean()

0.15286624203821655

Here, we can see that a negative $t$-statistic means that a greater percentage of Democrats voted 'yes' on the bill than Republicans. Therefore...

2. On the issue "aid-to-contras", Democrats support the issue more than Republicans, with a $t$-statistic of -18.05 and a $p$-value of $2.82 \times 10^{-54} << 0.01$.

3. On the issue "crime", Republicans support the issue more than Democrats, with a $t$-statistic of 16.34 and a $p$-value of $9.95 \times 10^{-47} << 0.01$.

4. The issue "water-project" is the only issue that satisfies the criteria. Republicans and Democrats support the issue equally, since we see a $t$-statistic of 0.09 and a $p$-value of $0.93 > 0.10$.
